File: backend/app/schemas/topology.py
"""Topology API contracts (M2-10): response models for ``/api/v1/topology``.

Pure data (D2): validation/serialization only, no I/O.  The graph endpoint
returns the projected Neo4j subgraph (:class:`GraphResponse`) and the diff
endpoint returns the M2-08 :class:`~app.engines.topology.diff.TopologyDiff`
wire shape wrapped with the two run ids it was computed from
(:class:`TopologyDiffResponse`).

Every graph response surfaces ``projected_at`` (ADR-0005: answers are "as of
run X") — the most recent ``last_projected_at`` stamp across the returned
nodes, or ``None`` when the subgraph is empty.
"""

from __future__ import annotations

import uuid
from typing import Any

from pydantic import BaseModel, ConfigDict, Field

from app.engines.topology.diff import TopologyDiff

__all__ = [
    "GraphEdge",
    "GraphNode",
    "GraphResponse",
    "ImpactDependency",
    "ImpactDependent",
    "ImpactResponse",
    "ImpactTarget",
    "TopologyDiffResponse",
]


class GraphNode(BaseModel):
    """One projected node: its label, key value, and flat property map."""

    model_config = ConfigDict(extra="forbid")

    label: str = Field(description="Projected node label (e.g. Device, Subnet).")
    key: Any = Field(description="Value of the label's key property (pg_id / natural key).")
    properties: dict[str, Any] = Field(
        default_factory=dict, description="Flat, JSON-safe property map for display."
    )


class GraphEdge(BaseModel):
    """One projected relationship between two node keys."""

    model_config = ConfigDict(extra="forbid")

    type: str = Field(description="Relationship type (e.g. CONNECTED_TO, ROUTES_TO).")
    source: Any = Field(description="Key of the start node.")
    target: Any = Field(description="Key of the end node.")
    properties: dict[str, Any] = Field(
        default_factory=dict, description="Flat, JSON-safe relationship property map."
    )


class GraphResponse(BaseModel):
    """The projected topology subgraph for ``GET /topology/graph``.

    ``projected_at`` is the most recent ``last_projected_at`` (ISO-8601 UTC)
    across the returned nodes — the projection pass these answers are "as of".
    It is ``None`` only when the filtered subgraph contains no nodes.
    """

    model_config = ConfigDict(extra="forbid")

    nodes: list[GraphNode] = Field(default_factory=list)
    edges: list[GraphEdge] = Field(default_factory=list)
    projected_at: str | None = Field(
        default=None,
        description="Most recent projection timestamp across the returned nodes.",
    )


class ImpactTarget(BaseModel):
    """A node addressed by label + key (the impact endpoint's minimal endpoint form)."""

    model_config = ConfigDict(extra="forbid")

    label: str = Field(description="Projected node label (e.g. Device, IPAddress, Application).")
    key: Any = Field(description="Value of the label's key property (pg_id / natural key).")


class ImpactDependent(BaseModel):
    """One application impacted by the target, with the edge's provenance (§8).

    ``application`` is the dependent application node; ``target`` is the endpoint
    it depends on (the target itself for a direct dependent, or a neighborhood
    node for an indirect one). ``sources``/``provenance``/``derived_at`` are the
    compact, refs-only explainability summary carried on the ``DEPENDS_ON`` edge.
    """

    model_config = ConfigDict(extra="forbid")

    application: GraphNode
    target: ImpactTarget
    sources: list[str] = Field(default_factory=list)
    provenance: list[str] = Field(default_factory=list)
    derived_at: str | None = None


class ImpactDependency(BaseModel):
    """One endpoint an Application target depends on, with the edge's provenance."""

    model_config = ConfigDict(extra="forbid")

    target: GraphNode
    sources: list[str] = Field(default_factory=list)
    provenance: list[str] = Field(default_factory=list)
    derived_at: str | None = None


class ImpactResponse(BaseModel):
    """The bounded impact answer for ``GET /topology/impact`` (ADR-0052 §8).

    ``dependents`` answers "what depends on X" (every target kind); for an
    ``Application`` target ``dependencies`` also answers "what does A depend on".
    ``projected_at`` is the watermark these answers are "as of" (``None`` when
    the target is absent from the projection); ``depth_used`` is the hop bound
    the physical-neighborhood expansion ran with.
    """

    model_config = ConfigDict(extra="forbid")

    target: ImpactTarget
    dependents: list[ImpactDependent] = Field(default_factory=list)
    dependencies: list[ImpactDependency] = Field(default_factory=list)
    projected_at: str | None = None
    depth_used: int


class TopologyDiffResponse(BaseModel):
    """The diff between two topology snapshots for ``GET /topology/diff``.

    Wraps the M2-08 :class:`TopologyDiff` (the four added/removed lists) with
    the two run ids the diff was computed from.
    """

    model_config = ConfigDict(extra="forbid")

    from_run: uuid.UUID = Field(description="The earlier run whose snapshot is the baseline.")
    to_run: uuid.UUID = Field(description="The later run whose snapshot is compared.")
    diff: TopologyDiff

File: frontend/src/api/topology.ts
/**
 * Typed client functions for the topology endpoints (M2-12).
 *
 * Mirrors the backend schemas in ``app/schemas/topology.py`` and
 * ``app/engines/topology/diff.py``, and the routes in
 * ``app/api/v1/topology.py`` (M2-10).
 */

import { apiFetch } from "./client";

// ── Node / Edge types ─────────────────────────────────────────────────────────

/**
 * One projected node: its label, key value, and flat property map.
 *
 * Mirrors ``GraphNode`` in ``app/schemas/topology.py``.
 */
export interface TopologyNode {
  /** Projected node label, e.g. ``Device``, ``Subnet``. */
  label: string;
  /** Value of the label's key property (pg_id UUID or natural key string). */
  key: string;
  /** Flat, JSON-safe property map for display. */
  properties: Record<string, unknown>;
}

/**
 * One projected relationship between two node keys.
 *
 * Mirrors ``GraphEdge`` in ``app/schemas/topology.py``.
 */
export interface TopologyEdge {
  /** Relationship type, e.g. ``CONNECTED_TO``, ``ROUTES_TO``. */
  type: string;
  /** Key of the start node. */
  source: string;
  /** Key of the end node. */
  target: string;
  /** Flat, JSON-safe relationship property map. */
  properties: Record<string, unknown>;
}

// ── Response shapes ───────────────────────────────────────────────────────────

/**
 * The projected topology subgraph returned by ``GET /topology/graph``.
 *
 * Mirrors ``GraphResponse`` in ``app/schemas/topology.py``.
 * ``projected_at`` is ``null`` when the filtered subgraph contains no nodes.
 */
export interface TopologyGraph {
  nodes: TopologyNode[];
  edges: TopologyEdge[];
  /** ISO-8601 UTC timestamp of the most recent projection pass, or ``null``. */
  projected_at: string | null;
}

/**
 * The diff result from M2-08 between two topology snapshots.
 *
 * Each field is a lexicographically sorted list in the canonical element form:
 * - ``nodes_added`` / ``nodes_removed``: ``[[label, key], ...]``
 * - ``edges_added`` / ``edges_removed``: ``[[rel_type, src_key, dst_key], ...]``
 *
 * Mirrors ``TopologyDiff`` in ``app/engines/topology/diff.py``.
 */
export interface TopologyDiff {
  /** Nodes present in ``to_run`` but absent in ``from_run``. */
  nodes_added: string[][];
  /** Nodes present in ``from_run`` but absent in ``to_run``. */
  nodes_removed: string[][];
  /** Edges present in ``to_run`` but absent in ``from_run``. */
  edges_added: string[][];
  /** Edges present in ``from_run`` but absent in ``to_run``. */
  edges_removed: string[][];
}

/**
 * The diff between two topology snapshots returned by ``GET /topology/diff``.
 *
 * Wraps {@link TopologyDiff} with the two run ids the diff was computed from.
 * Mirrors ``TopologyDiffResponse`` in ``app/schemas/topology.py``.
 */
export interface TopologyDiffResponse {
  /** The earlier (baseline) run id. */
  from_run: string;
  /** The later (compared) run id. */
  to_run: string;
  diff: TopologyDiff;
}

// ── Query-string params ───────────────────────────────────────────────────────

/** Upper bound on the neighborhood ``depth`` (mirrors the backend constant). */
export const MAX_NEIGHBORHOOD_DEPTH = 5;

/** Optional filters for ``GET /topology/graph``. */
export interface TopologyGraphParams {
  /** Scope to devices assigned to this site name. */
  site?: string;
  /** Scope to nodes belonging to this VRF. */
  vrf?: string;
  /**
   * Relationship families to include in the response.
   * ``l2`` — LLDP/CDP neighbor links only.
   * ``l3`` — subnet adjacency and routing links only.
   * ``dns`` — DNS zone/record dependency layer (T13 DnsZone/DnsRecord/RESOLVES_TO).
   * ``app`` — application-dependency layer (P4 W2 Application/DEPENDS_ON).
   * ``all`` — all relationship types (default server-side).
   */
  layer?: "l2" | "l3" | "dns" | "app" | "all";
}

/** Params for ``GET /topology/graph/neighborhood`` (audit Wave 5, scoped read). */
export interface TopologyNeighborhoodParams {
  /** Key (``pg_id``) of the device at the center of the neighborhood. */
  device: string;
  /** Hop radius, 1..{@link MAX_NEIGHBORHOOD_DEPTH} (server default 2). */
  depth?: number;
  /** Relationship families to walk — same semantics as the graph read. */
  layer?: TopologyGraphParams["layer"];
}

// ── API functions ─────────────────────────────────────────────────────────────

/**
 * ``GET /api/v1/topology/graph`` — return the projected Neo4j subgraph.
 *
 * @param params - Optional filters: ``site``, ``vrf``, ``layer``.
 * @returns The projected topology graph as of the latest projection pass.
 * @throws {ApiError} For any non-2xx response (RFC 7807 problem document).
 */
export function getTopologyGraph(params: TopologyGraphParams = {}, signal?: AbortSignal): Promise<TopologyGraph> {
  const qs = new URLSearchParams();
  if (params.site !== undefined) qs.set("site", params.site);
  if (params.vrf !== undefined) qs.set("vrf", params.vrf);
  if (params.layer !== undefined) qs.set("layer", params.layer);
  const query = qs.toString();
  return apiFetch<TopologyGraph>(`/topology/graph${query ? `?${query}` : ""}`, { signal });
}

/**
 * ``GET /api/v1/topology/graph/neighborhood`` — the subgraph within ``depth``
 * hops of one device (the scoped read the topology page defaults to).
 *
 * @param params - ``device`` (required), optional ``depth`` and ``layer``.
 * @returns The device-centered neighborhood subgraph.
 * @throws {ApiError} 404 when the device key is not in the projection; 422 for
 *   an out-of-range ``depth``.
 */
export function getTopologyNeighborhood(
  params: TopologyNeighborhoodParams,
  signal?: AbortSignal,
): Promise<TopologyGraph> {
  const qs = new URLSearchParams({ device: params.device });
  if (params.depth !== undefined) qs.set("depth", String(params.depth));
  if (params.layer !== undefined) qs.set("layer", params.layer);
  return apiFetch<TopologyGraph>(`/topology/graph/neighborhood?${qs.toString()}`, { signal });
}

/**
 * ``GET /api/v1/topology/diff`` — diff the snapshots of two discovery runs.
 *
 * @param fromRun - The earlier (baseline) run id (UUID string).
 * @param toRun - The later (compared) run id (UUID string).
 * @returns Added and removed nodes and edges between the two run snapshots.
 * @throws {ApiError} 404 when either run has no snapshot; 422 for invalid UUIDs.
 */
export function getTopologyDiff(fromRun: string, toRun: string, signal?: AbortSignal): Promise<TopologyDiffResponse> {
  const qs = new URLSearchParams({ from_run: fromRun, to_run: toRun });
  return apiFetch<TopologyDiffResponse>(`/topology/diff?${qs.toString()}`, { signal });
}
